rero_ils/modules/acquisition/rollover.py

```diff
@@ -0,0 +1,664 @@
+        :param original_budget: the `Budget` resource related to resources to
+            migrate.
+        :param destination_budget: the `Budget` resource where the rollovered
+            resources will be migrated.
+        :param logging_config: (optional) a dictionary containing all necessary
+            configuration to log the rollover process result. If not specified
+            the configuration comes from `ROLLOVER_LOGGING_CONFIG` setting.
+        :param is_interactive: boolean to determine if user confirmation is
+            required. True by default.
+        :param propagate_errors: Boolean to determine if error will be
+            propagated to caller. Default is false
+        :param **kwargs: all others named argument useful for rollover process.
+        :raises InactiveBudgetException: if ....
+        """
+        self.cache = {}
+        self.stack = []
+        self.mapping_table = {}
+        self.propagate_errors = propagate_errors
+
+        # Set special logging configuration for rollover process
+        logging.config.dictConfig(logging_config or ROLLOVER_LOGGING_CONFIG)
+        self.logger = logging.getLogger(__name__)
+        self.logger.info('ROLLOVER PROCESS ==================================')
+
+        self.original_budget = original_budget
+        if destination_budget is None:
+            destination_budget = self._create_new_budget(**kwargs)
+
+        self.destination_budget = destination_budget
+        self.is_interactive = is_interactive
+        # Why `_validate` is into a try...except.
+        #   It's possible that the stack already contains an object if the
+        #   destination budget didn't exist and has been created. In this case
+        #   this destination budget must be destroyed.
+        try:
+            self._validate()
+        except Exception as e:
+            self._abort_rollover(str(e))
+            if self.propagate_errors:
+                raise e
+
+    def run(self):
+        """Run the rollover process."""
+        log = self.logger
+        try:
+            log.info('start running....')
+            log.info("parameters ::")
+            log.info(f"\torigin      : {self.original_budget}")
+            log.info(f"\tdestination : {self.destination_budget}")
+
+            # STEP#1 :: Get account related to original budget
+            #   Load acquisition accounts related to the original budget and
+            #   sort them to get a hierarchical tree structure (parent -> child
+            #   -> grandchild -> ...).
+            #   When loaded, log these accounts into a table to visualize which
+            #   resources will be migrated according to related library
+            #   rollover setting.
+            orig_accounts = AcqRollover._get_accounts(self.original_budget.pid)
+            if not orig_accounts:
+                raise RolloverError('Unable to find any account to rollovered')
+            log.info("original accounts ::")
+            columns = [
```

Comment:
Is this can be in the constructor?

Reply:
not really because we display two tables during the script process. `columns` variable are redefined at line#252